specs/withdrawals.md

```diff
@@ -0,0 +1,176 @@
+| `WithdrawalMessageInclusionProof` | `messageProof`     |
+
+These inputs must satisfy the following conditions:
+
+1. The `timestamp` is at least `FINALIZATION_WINDOW` seconds old.
+1. `OutputOracle.l2Outputs(timestamp)` returns a non-zero value `l2Output`.
+1. The `storageRootProof` is valid for the `l2Output` returned in step 2, according to the SSZ encoding described in the
+   [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+1. The `messageProof` is valid for the provided `message` and `messageProof`
+
+## Withdrawal Flow
+
+**On L2:**
+
+1. An L2 account sends a withdrawal message (and possibly also ETH) to the `Withdrawoooor` predeploy contract.
+   This is a very simple contract that stores a mapping from the hash of the `WithdrawalMessage` as defined above to a
+   boolean value. (`mapping (bytes32 => bool) withdrawalMessages`)
+2. If ETH is being withdrawn, it can eventually be burned by deploying a contract which immediately `SELFDESTRUCT`s.
+   This has the benefit of using an existing EVM mechanism for removing the ETH from the world state, without having to
+   add a diff to L2 Geth.
+
+**On L1:**
+
+1. A [relayer][g-relayer] submits the required inputs to the `DepositFeed` contract. The relayer may or may not be the same entity which
+   initiated the withdrawal on L2.
+2. The `DepositFeed` contract retrieves the output root from the `OutputOracle`'s `l2Outputs()` function, and performs
+   the remainder of the verification process internally.
+3. If verification is successful, the message is forwarded to the target.
+    1. If the message call is successful, the hash is stored in a `successfulMessages` mapping.
+    2. Otherwise it is stored in a `repeatableMessages` mapping.
+4. If verification fails, the call reverts.
+
+## The L2 Withdrawals Contract
+
+The L2 Withdrawals predeploy is a simple contract at `0x4200000000000000000000000000000000000015` which stores messages
+to be withdrawn.
+
+It contains a mapping which records withdrawals.
+
+```js
+interface Withdrawoooor {
+
+    event WithdrawalMessage(
+        address indexed target,
+        address sender,
+        bytes message,
+        uint256 value,
+        uint256 messageNonce, // this is a global nonce value for all withdrawal messages
+        uint256 gasLimit
+    );
+
+    function withdraw(
+        address target,
+        address sender,
+        bytes message,
+        uint256 gasLimit
+    ) external payable;
+
+    function burn();
+}
+```
+
+## Security Considerations
+
+### Key Properties of Withdrawal Verification
+
+1. It should not be possible 'double spend' a withdrawal, ie. to relay a `WithdrawalMessage` on L1 which does not correspond to a message
+  initiated on L2. For reference, see [this writeup][polygon-dbl-spend] of a vulnerability of this type found on Polygon.
+
+    [polygon-dbl-spend]: https://gerhard-wagner.medium.com/double-spending-bug-in-polygons-plasma-bridge-2e0954ccadf1
+
```

Comment:
Addittionally (can be a separate bullet), it should not be possible to relay the same message multiple times.